### ISSUE TITLE
Warning when building the documentation

### ISSUE THREAD

**Issue by enadeau** [stravalib/stravalib#375]
When trying to build the documentation, three warnings are raised and they all translate to minore issue in the doc display.

- [x] [XRateLimiteRule `__init__` function](https://stravalib.readthedocs.io/en/latest/reference/api/stravalib.util.limiter.XRateLimitRule.html) the description of the force limits param ends up on an other line

- [x] [Client.upload_activity](https://stravalib.readthedocs.io/en/latest/reference/api/stravalib.client.Client.upload_activity.html) the display message of description for the deprecated param is not in the deprecated box
- [ ] [BaseEffort](https://stravalib.readthedocs.io/en/latest/reference/model.html#stravalib.model.BaseEffort) model json schema is not computed correctly for the activity field.

I have a pull request ready for first and second error but I'm not quite sure about 3 so I decided to open an issue about it.

**Comment by lwasser** [comment]
@enadeau i didn't follow this PR closely. did the merged PR #376 address this issue so we can close it? many thanks.

**Comment by enadeau** [comment]
@lwasser the third issue is not addressed

**Comment by lwasser** [comment]
ok i dug into this just a bit more last night. I DID see the warning originally when building the docs  in [your first typing pr](https://github.com/stravalib/stravalib/pull/374) .
However, when i built again last night against master with the two other pr's merged, that warning seemed to be gone. 

I'm curious - do you still see it when building the docs?

**Comment by enadeau** [comment]
Yes I still see it when building the doc on master
```
reading sources... [100%] reference/utilities                                                            
/home/enadeau/Documents/stravalib/docs/autodoc_pydantic.rst: WARNING: JSON schema can't be generated for 'stravalib.model.BaseEffort' because the following pydantic fields can't be serialized properly: ['activity'].
```

**Comment by lwasser** [comment]
ok interesting. 

my pr just also [yielded a pydantic related build error](https://github.com/stravalib/stravalib/actions/runs/5425675102/jobs/9866738424#step:8:21)

and that build is using [pydantic 2.0](https://github.com/stravalib/stravalib/actions/runs/5425675102/jobs/9866738424#step:7:19)

BUT when i build locally i do NOT see that warning. But I am running pydantic 1.10.9 as i'm a conda user. and they JUST (1 hour ago) [released 2.0. ](https://github.com/pydantic/pydantic/releases/tag/v2.0) - which means pydantic on conda is likely on its way but not merged yet. 

what version of pydantic are you on? i did have to force an update to ensure things ran properly for me the other night. 

It looks like we might have to either:
*  implement some more updates given the latest version or 
* pin pydantic until we are ready to implement updates. 
 
so this may or may not be related to what you are seeing but i suspect (but could be wrong!!) that the pydantic version might be at the root if this all. 🤔 🔍

**Comment by lwasser** [comment]
i am incorrect - that warning is STILL there in the doc build when pinned to 1.10.9. i have opened another issue for the pydantic 2.0 release and we can continue to explore this issue here. my apologies for conflating two different bugs!!

**Comment by enadeau** [comment]
If you want to try reproducing or maybe just compare with your environment, here's the environment I get the bug in running in 3.11.4
```
alabaster==0.7.13
arrow==1.2.3
asttokens==2.2.1
attrs==23.1.0
autodoc-pydantic==1.8.0
Babel==2.12.1
backcall==0.2.0
beautifulsoup4==4.12.2
certifi==2023.5.7
cfgv==3.3.1
charset-normalizer==3.1.0
click==8.1.3
colorama==0.4.6
comm==0.1.3
contourpy==1.1.0
coverage==7.2.7
cycler==0.11.0
debugpy==1.6.7
decorator==5.1.1
distlib==0.3.6
docutils==0.19
executing==1.2.0
fastjsonschema==2.17.1
filelock==3.12.2
fonttools==4.40.0
greenlet==2.0.2
identify==2.5.24
idna==3.4
imagesize==1.4.1
importlib-metadata==6.7.0
iniconfig==2.0.0
ipykernel==6.23.3
ipython==8.14.0
jedi==0.18.2
Jinja2==3.1.2
jsonschema==4.17.3
jupyter-cache==0.6.1
jupyter_client==8.3.0
jupyter_core==5.3.1
kiwisolver==1.4.4
livereload==2.6.3
markdown-it-py==2.2.0
MarkupSafe==2.1.3
matplotlib==3.7.1
matplotlib-inline==0.1.6
mdit-py-plugins==0.3.5
mdurl==0.1.2
mypy==1.4.0
mypy-extensions==1.0.0
myst-nb==0.17.2
myst-parser==0.18.1
nbclient==0.7.4
nbformat==5.9.0
nest-asyncio==1.5.6
nodeenv==1.8.0
numpy==1.25.0
packaging==23.1
parso==0.8.3
pexpect==4.8.0
pickleshare==0.7.5
Pillow==9.5.0
Pint==0.22
platformdirs==3.8.0
pluggy==1.2.0
pre-commit==3.3.3
prompt-toolkit==3.0.38
psutil==5.9.5
ptyprocess==0.7.0
pure-eval==0.2.2
pydantic==1.10.9
pydata-sphinx-theme==0.12.0
Pygments==2.15.1
pyparsing==3.1.0
pyright==1.1.315
pyrsistent==0.19.3
pytest==7.4.0
pytest-cov==4.1.0
python-dateutil==2.8.2
pytz==2023.3
pyupgrade==3.7.0
PyYAML==6.0
pyzmq==25.1.0
requests==2.31.0
responses==0.23.1
six==1.16.0
snowballstemmer==2.2.0
soupsieve==2.4.1
Sphinx==5.3.0
sphinx-autobuild==2021.3.14
sphinx-autodoc-typehints==1.23.3
sphinx-copybutton==0.5.2
sphinx-remove-toctrees==0.0.3
sphinx_design==0.4.1
sphinx_inline_tabs==2023.4.21
sphinxcontrib-applehelp==1.0.4
sphinxcontrib-devhelp==1.0.2
sphinxcontrib-htmlhelp==2.0.1
sphinxcontrib-jsmath==1.0.1
sphinxcontrib-qthelp==1.0.3
sphinxcontrib-serializinghtml==1.1.5
sphinxext-opengraph==0.8.2
SQLAlchemy==2.0.17
stack-data==0.6.2
-e git+ssh://git@github.com/enadeau/stravalib.git@dffb9284e56fd344446b9d54ec3ccad157d060be#egg=stravalib
tabulate==0.9.0
tokenize-rt==5.1.0
tornado==6.3.2
traitlets==5.9.0
types-click==7.1.8
types-Flask==1.1.6
types-Jinja2==2.11.9
types-MarkupSafe==1.1.10
types-pytz==2023.3.0.0
types-PyYAML==6.0.12.10
types-requests==2.31.0.1
types-urllib3==1.26.25.13
types-Werkzeug==1.0.9
typing_extensions==4.6.3
urllib3==2.0.3
virtualenv==20.23.1
wcwidth==0.2.6
zipp==3.15.0
```

**Comment by lwasser** [comment]
thank you for this!! so i've figured this out (i think)

1. numerous classes (and a few functions) are not rendering in the docs as they are missing simple docstrings. so i'm fixing that in model.py now
2. I also noticed that BestEffort has NO content in the body of the object. it inherits from BaseEffort. But i believe that JSON issue is related to the fact that there is no body in the class. i have added a `pass` and it renders properly

Without that BestEffort JSON renders oddly with no title, and initial description and numerous `$ref` values throughout. It seems like the pydantic autodoc module is sensitive to docstrings being present and somehow gets mixed up when things are missing.

because i'm a bit nervous about breaking things with a class in an object within a library that has a complex inheritance structure i'm going to open a PR once i've fixed the entire module and would appreciate a review. But i do think the path to resolving this warning is clear at this point! And it is NOT an environment issue as i thought (i had thought perhaps a version bug given my initial experiences with mypy)

more once the PR is opened!!

**Comment by lwasser** [comment]
hey y'all - i looked more into this and [opened this issue](https://github.com/mansenfranzen/autodoc_pydantic/issues/157). It seems like this is an issue with the `to_json()` method for this specific class. @jsamoocha we could configure [autodoc-pydantic](https://autodoc-pydantic.readthedocs.io/en/stable/users/configuration.html#show-schema-json-error-strategy) to not raise this warning. but it also renders awkwardly in the docs as you can see below.


![image](https://github.com/stravalib/stravalib/assets/7649194/62f996ea-3915-4b79-bd3f-37ca65da77ca)

**Comment by jsamoocha** [comment]
Thanks for researching this. I'm now seeing the potential issue: there's a circular reference `Activity->list[BestEffort]->Activity`. This can lead to infinite recursion problems when generating JSON (only in some edge cases, possibly including the documentation).

This situation is a consequence of backward compatibility with pre-1.2 versions, where there was a single `Activity` model as opposed to Strava's hierarchy of `Meta`-, `Summary`- and `DetailedActivity`.

I'm unsure how to work around this in the short term.

**Comment by lwasser** [comment]
ok so we can simply ignore the warning for now OR maybe turn it off. But it seems like a good warning to know about. should I just leave that as is?

I also know what's causing that findfonts warning now. it's a known issue with matplotlib which is an (odd) dependency of a dependency associated with the [openGraph](https://github.com/wpilibsuite/sphinxext-opengraph) which allows us to populate better metadata in our package and show cards if shared on social search etc. most people silence it by simply installing the missing font into the ci build. that should be a quick fix once i have time to implement it.

**Comment by jsamoocha** [comment]
Let's fix the type hierarchy (i.e., introduce Strava's meta-, summary-, and detailed- types) whenever we introduce other breaking changes, e.g. when moving to a 2.0 release. We can leave this warning as-is. It doesn't seem to lead to spectacular rendering issues of the docs.

Do you think we can close this issue?

**Comment by lwasser** [comment]
I think we can close it if the plan is go fix the hierarchy before 2.0
release.

I'll work on the fonts issue when I have some time after travel.
